geo/src/main/java/org/apache/carbondata/geo/GeoHashImpl.java

```diff
@@ -0,0 +1,400 @@
+    this.userDefineMinLongitude = Double.valueOf(minLongitude);
+    this.userDefineMinLatitude = Double.valueOf(minLatitude);
+    this.gridSize = Integer.parseInt(gridSize);
+    this.conversionRatio = Integer.parseInt(conversionRatio);
+    this.lon0ByRation = userDefineMinLongitude * this.conversionRatio;
+    this.lat0ByRation = userDefineMinLatitude * this.conversionRatio;
+    calculateData();
+  }
+
+  /**
+   * Generates the GeoHash ID column value from the given source columns.
+   * @param sources Longitude and Latitude
+   * @return Returns the generated hash id
+   * @throws Exception
+   */
+  @Override
+  public String generate(List<?> sources) throws Exception {
+    if (sources.size() != 2) {
+      throw new RuntimeException("Source columns list must be of size 2.");
+    }
+    if (sources.get(0) == null || sources.get(1) == null) {
+      // Bad record. Just return null
+      return null;
+    }
+    if (!(sources.get(0) instanceof Long) || !(sources.get(1) instanceof Long)) {
+      throw new RuntimeException("Source columns must be of Long type.");
+    }
+    Long longitude = (Long) sources.get(0);
+    Long latitude  = (Long) sources.get(1);
+    // generate the hash code
+    int[] gridPoint = calculateID(longitude, latitude);
+    Long hashId = createHashID(gridPoint[0], gridPoint[1]);
+    return String.valueOf(hashId);
+  }
+
+  /**
+   * Query processor for GeoHash.
+   * example: POLYGON (35 10, 45 45, 15 40, 10 20, 35 10)
+   * so there will be a sample check
+   * @param polygon a group of pints, close out to form an area
+   * @return Returns list of ranges of GeoHash IDs
+   * @throws Exception
+   */
+  @Override
+  public List<Long[]> query(String polygon) throws Exception {
+    if (!validate(polygon)) {
+      return null;
+    } else {
+      String[] pointList = polygon.trim().split(",");
+      List<double[]> queryList = new ArrayList<>();
+      for (String str: pointList) {
+        String[] points = str.split(" ");
+        if (2 != points.length) {
+          throw new RuntimeException("longitude and latitude is a pair need 2 data");
+        } else {
+          try {
+            queryList.add(new double[] {Double.valueOf(points[0]), Double.valueOf(points[1])});
+          } catch (NumberFormatException e) {
+            throw new RuntimeException("can not covert the string data to double", e);
+          }
+        }
+      }
+      List<Long[]> rangeList = getPolygonRangeList(queryList);
+      return rangeList;
+    }
+  }
+
+  private boolean validate(String polygon) throws Exception{
+    String[] pointList = polygon.trim().split(",");
+    if (4 > pointList.length) {
+      throw new RuntimeException("polygon need at least 3 points, really has " + pointList.length);
+    } else {
+      String firstPoint = pointList[0].trim();
+      String lastPoint = pointList[pointList.length - 1].trim();
+      if (!firstPoint.equals(lastPoint)) {
```

Comment:
Please split and check the long and lat values seperately. `validate()` can fail if the number of spaces between long and lat do not match between first and last points.

Reply:
done